### ISSUE TITLE
SAW script takes several minutes parsing Cryptol code on load

### ISSUE THREAD

**Issue by mrogers67** [GaloisInc/saw-script#1902]
This (reduced) SAW scripts takes several minutes to load/parse the following Cryptol code, which is before the interpreter executes subsequent commands. Everything works fine once parsing finishes but the load time is rather excessive.

```
let {{
    ex : [384 + 32] -> [384 + 32]
    ex x = zero # (take`{32} (drop`{352} x))

    xi  x = (drop`{384} x) # zero
    x0  x = (zero # (xi  x)) + (ex (xi  x))
    x1  x = (zero # (x0  x)) + (ex (x0  x))
    x2  x = (zero # (x1  x)) + (ex (x1  x))
    x3  x = (zero # (x2  x)) + (ex (x2  x))
    x4  x = (zero # (x3  x)) + (ex (x3  x))
    x5  x = (zero # (x4  x)) + (ex (x4  x))
    x6  x = (zero # (x5  x)) + (ex (x5  x))
    x7  x = (zero # (x6  x)) + (ex (x6  x))
    x8  x = (zero # (x7  x)) + (ex (x7  x))
    x9  x = (zero # (x8  x)) + (ex (x8  x))
    x10 x = (zero # (x9  x)) + (ex (x9  x))
    x11 x = (zero # (x10 x)) + (ex (x10 x))
}};
```

FWIW, removing `x9`...`x11` reduces the load time to ~10s.

**Comment by mrogers67** [comment]
See also #1901, which I encountered when trying to further reduce the above example.

**Comment by RyanGlScott** [comment]
The repeated use of `take` and `drop` (via `ex`) is causing each of the `x<N>` functions to have an exponentially larger number of variables in its inferred type. If you load this into Cryptol, you can see this:

```
Main> :t x0
x0 :
  {n, m, i, j}
    (m + (i + j) == 416, 416 == j + n, fin j, fin i, fin m) =>
    [384 + j] -> [m + (i + j)]
Main> :t x1
x1 :
  {n, m, i, j, k, n1, m1, i1}
    (i + (i1 + j) == 416, n1 + (m1 + i1) == 416, k + (n1 + (m1 + i1)) == 416,
     416 == i + (i1 + j), 416 == i1 + m, 416 == n + i1, fin j, fin i, fin i1,
     fin m1, fin n1, fin k) =>
    [384 + i1] -> [k + (n1 + (m1 + i1))]
Main> :t x2
x2 :
  {n, m, i, j, k, n1, m1, i1, j1, k1, n2, m2, i2, j2, k2, n3}
    (k1 + (n3 + n2) == 416, j1 + (k1 + (n3 + n2)) == 416, m1 + (n3 + i1) == 416,
     j2 + (k2 + n3) == 416, i2 + (j2 + (k2 + n3)) == 416, i + (j + n3) == 416,
     m2 + (i2 + (j2 + (k2 + n3))) == 416, 416 == j1 + (k1 + (n3 + n2)),
     416 == m1 + (n3 + i1), 416 == n3 + n1, 416 == n3 + k, 416 == i + (j + n3),
     416 == m + n3, 416 == n + n3, fin n2, fin k1, fin j1, fin i1, fin m1,
     fin n3, fin k2, fin j2, fin i2, fin j, fin i, fin m2) =>
    [384 + n3] -> [m2 + (i2 + (j2 + (k2 + n3)))]
...
```

As such, I don't think loading this file will ever be blazingly fast, considering the exponential growth that is occurring. (For example, `x11` has 8,192 variables in its type signature!) Still, perhaps there is some work that could be done to reduce some large constant factors on SAW's side. Loading this program into Cryptol isn't exactly fast either, but it does complete in a manner of seconds rather than minutes.

**Comment by sauclovian-g** [comment]
Well, also, we could avoid generating vacuous type variables (e.g. in ```x1```, ```k``` is always 0 and ```m``` is the same as ```n```) as well as redundant conditions (note ```i + (i1 + j) == 416``` and ```416 == i + (i1 + j)``` in ```x1``` and this would cut down the size of the blowup.

It appears from mucking with it a bit that the real number of type variables required is two for x0, three for x1, five for x2, nine for x3, etc., because the argument always goes through ```xi``` first and that introduces one, and each use of ```x0``` introduces one (that determines the width of its ```zero``` term) and there are twice as many uses of ```x0``` in each layer. So, ```x11``` really "only" needs 2049 variables...

(also, it seems like the fact that the ```zero``` term for ```x1``` and up, and thus the LHS of ```#```, is always width 0 should generate warnings.)

**Comment by RyanGlScott** [comment]
I agree that it would be nice if Cryptol could warn you if a type variable of kind `#` can only ever be equal to one particular number (such as `0` in the example above). This might require running an SMT solver in order to deduce, but since we are _already_ running an SMT solver during Cryptol's typechecker, the performance overhead of checking this might be reasonable.
